<a href="https://colab.research.google.com/github/boscolio/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [32]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

!wget  'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'

--2020-08-10 17:59:49--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data’

imports-85.data     100%[===================>]  25.33K  --.-KB/s    in 0.07s   

2020-08-10 17:59:49 (381 KB/s) - ‘imports-85.data’ saved [25936/25936]



In [35]:
# Creating the headers list
headers = ['symboling','norm_loss','make','fuel','aspiration','doors',
           'bod_style','drv_wheels','eng_loc','wheel_base','length','width',
           'height','curb_weight','engine','cylinders','engine_size',
           'fuel_system','bore','stroke','compression','hp','peak_rpm',
           'city_mpg','hgwy_mpg','price']

# Importing file as a dataframe with headers
df = pd.read_csv('imports-85.data', names=headers)

df.head()

symboling norm_loss         make fuel  ... peak_rpm city_mpg hgwy_mpg  price
0          3         ?  alfa-romero  gas  ...     5000       21       27  13495
1          3         ?  alfa-romero  gas  ...     5000       21       27  16500
2          1         ?  alfa-romero  gas  ...     5000       19       26  16500
3          2       164         audi  gas  ...     5500       24       30  13950
4          2       164         audi  gas  ...     5500       18       22  17450

[5 rows x 26 columns]

In [44]:
# Replacing '?' with NaN values
df = df.replace('?',np.NaN)
print(df.shape)
df.head()

(205, 26)


symboling norm_loss         make fuel  ... peak_rpm city_mpg hgwy_mpg  price
0          3       NaN  alfa-romero  gas  ...     5000       21       27  13495
1          3       NaN  alfa-romero  gas  ...     5000       21       27  16500
2          1       NaN  alfa-romero  gas  ...     5000       19       26  16500
3          2       164         audi  gas  ...     5500       24       30  13950
4          2       164         audi  gas  ...     5500       18       22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [40]:
# 10 cars seems like a small sample size given how varried highway mpg can be.
sample = df.sample(10, random_state=30)
sample['hgwy_mpg'].mean()

32.8

In [48]:
# My null hypothesis is that my sample mean is equal to my population mean 
# My alternative hypothesis is that my sample mean is not equal to my population mean
# My confidence is 95%
from scipy import stats
stats.ttest_1samp(df['hgwy_mpg'], 32.8, nan_policy='omit')

Ttest_1sampResult(statistic=-4.259684116248613, pvalue=3.124442782308114e-05)

In [49]:
# Due to a t-statistic of -4.25, and a p-value of .000031, we reject the null 
# hypothesis that the sample mean is equal to the population mean and suggest 
# the alternative, that they are not equal.

df['hgwy_mpg'].mean()

30.75121951219512

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [51]:
# The null hypothesis is that the cars have a highway mpg of 35
# The alternative hypothesis is that the cars do not have a highway mpg of 35
stats.ttest_1samp(sample['hgwy_mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

In [ ]:
# We reject the hypothesis that the cars have a higway mpg 0f 35 due to a 
# t-statistic of -0.96 and a p-value of .36

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [54]:
sample2 = df.sample(100, random_state=30)
stats.ttest_1samp(sample2['hgwy_mpg'], 35, nan_policy='omit')

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

In [ ]:
# We fail to reject the hypothesis that the cars have a highway mpg of 35 due to
# a t-statistic of -6.52 and a p-value of .000000003.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Our smaller sample did not accurately represent the highway mpg of these cars, where as the larger sample size was a better representation and better supported his claim.


## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [55]:
!wget 'https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data'

--2020-08-10 20:28:45--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-08-10 20:28:45 (277 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [63]:
# Assigning header to a list
headers= ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']
# Assigning dataframe
df = pd.read_csv('house-votes-84.data', names=headers)

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [64]:
# Replacing 'y', 'n' and '?'
df = df.replace({'y' : 1, 'n' : 0, '?' : np.NaN})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [65]:
# Separating the data by 'party'
dem = df[df['party'] == 'democrat']
rep = df[df['party'] == 'republican']

In [67]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [68]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [69]:
# Looking at the .head() outputs it seems 'budget' is a good place to start for the Democrats

# Null Hypothesis
# Democratic support of the budget bill is equal to Republican support of the budget bill 
# Alternative Hypothesis
# Democratic support of the budget bill is not equal to Republican support of the budget bill
# Confidence is 99%

stats.ttest_ind(dem['budget'], rep['budget'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

Based on a t-statistic of 23.2, and a p-value of 2.07e-77 , I fail to reject the null hypothesis that Democratic support is different from Republican levels of support for the budget.

Based on the t-statistic of 23.2 I can conclude Democratic support for the budget is higher than Republican support.

In [71]:
# Now we will compare support for the education bill

# Null Hypothesis
# Republican support of the education bill is equal to Democratic support of the education bill
# Alternative Hypothesis
# Republican support of the education bill is not equal to Democratic support of the education bill
# Confidence is 99%

stats.ttest_ind(rep['education'], dem['education'], nan_policy='omit')

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

Based on a t-statistic of 20.5, and a p-value of 1.88e-64, I fail to reject the null hypothesis that Republican support for the education bill is different from Democratic levels of support for the education bill.

Based on the t-statistic of 20.5 I can conclude Republican support for the education bill is higher than Republican support.

In [72]:
# Finally we will compare support for the "water project"

# Null Hypothesis
# Republican support of the water project bill is equal to Democratic support of the water project bill
# Alternative Hypothesis
# Republican support of the water project bill is not equal to Democratic support of the water project bill
# Confidence is 90%

stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

Based on a t-statistic of 0.089, and a p-value of 0.929, I reject the null hypothesis that Republican support for the water project bill is different from Democratic levels of support for the water project bill.

Based on the t-statistic of 0.089 I can conclude Republican support for the education bill is higher than Democratic support.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
for bill in 



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE